### DISCUSSION TITLE
테스트 코드 작성과 캡슐화의 균형

### DISCUSSION THREAD

**Discussion by fkdl0048** [fkdl0048/Mentoring-Project#110] [Q&A] [answered]
멘토링때 이야기 했지만.. 좀 더 정리하고 싶어서 글을 작성합니다.

테스트 코드를 작성 가능한 코드를 짜다 보면 기존 코딩 방식에서 조금 벗어나는 느낌이 듭니다.

MVP 패턴의 경우도 올바른 동작을 비교하기 위해 View에 반영된 데이터와 Presenter 데이터도 같이 오픈 해야 하는 상황이 오게 되는데 어떻게 접근해야 좋은 구조로 갈 수 있을까요?

테스트 코드를 작성하기 위해 클래스를 더 세분화하고 기능에 맞게 설계하는 점은 너무 좋지만, 테스트가 가능해야 해서 캡슐화가 깨지는 경우가 발생하는 것 같습니다.

아니면 get으로 해당 데이터만 반환하는 경우는 문제가 없는 걸까요?

**Comment by jongfeel** [comment, marked as answer]
데이터는 숨기고 메서드를 통해 공개하는게 일반적인 방식입니다. 그걸 캡슐화라고 하는 거죠.
데이터를 직접적으로 getter로 주는 게 번거롭다면 메서드의 반환 값으로 체크해도 충분합니다.

이해하기에 충분한 캡슐화 테스트 에제를 하나 들어보면
List에 Add 메서드를 검증하기 위해 indexer로 Add 메서드를 통해 추가한 데이터를 검증하는 방식이 맞다고 생각하고 테스트 코드를 짜보면

``` c#
[Test]
public void List_Add_Test()
{
    // Arrange
    List<int> list = new List<int>();

    // Act
    list.Add(1);

    // Assert
    Assert.AreEqual(1, list[0]);
}
```

그런데 public method 들 중에 다른 걸 활용하면 assert를 할 수 있습니다.
관점을 Add 메서드의 올바른 동작에 초점을 맞춘다면 캡슐화가 되어 있는 다른 public method로 검증해도 되는 거죠.

``` c#
[Test]
public void List_Add_Test()
{
    // Arrange
    List<int> list = new List<int>();

    // Act
    list.Add(1);

    // Assert
    Assert.Greater(list.Count(), 0);
}
```


객체가 하려는 행위를 데이터의 set, get이 아닌 행위를 통해 그 역할이 무엇인지 집중해 보면 좋은 메서드 설계가 됩니다.
그러면 인터페이스를 통한 객체지향의 설계 원칙에 다가서는 좋은 시점이 될 수도 있습니다.

만약 Rectangle이라는 클래스를 설계했을 때
new Rectangle(x1, y1, x2, y2); 로 생성해 놓고 x1, y1, x2, y2의 getter를 4개를 만드는 방법 보다는
Rectange이 하려는 행위와 그 결과를 생각해서 넓이를 구하는게 그 역할 중 하나라고 설계했다면
rectangle.Area() 메서드를 통해 확인할 수 있게 만들면 캡슐화를 잘 보여주는 설계라고 할 수 있겠죠.

팩트) 실제 C#의 Rectangle struct에는 Area() 메소드는 없습니다.

**Reply by fkdl0048** [reply]
말씀해주신 말이 뭔지 이해했습니다.

그냥 테스트 코드라고 해서 값을 비교하여 올바른 값이 출력된 것을 검증하는 것이 아닌 해당 객체의 목적에 맞춰 작성하는 것이 중요한 부분인 것 같습니다. (지금까지는 무식하게 작성한 것 같습니다 ㅎ..)

왜 인지 플레이어 관련 코드는 테스트 코드 짜는 데 무리가 없다고 느꼈는데, UI 관련 코드에서 자꾸 무식하게 테스트 코드를 짜려고 해서 혼란이 온 것 같네요.. 

말씀해주신대로 해당 객체의 목적을 제대로 파악하지 못하고 작업해서 그런 것 같습니다..

*여기서 말씀해주신 인터페이스가 그 인터페이스가 아닌게 맞나요? 제가 이해한 바로는 C#의 인터페이스가 아닌 사전적의미로 해석했습니다. 객체가 상호작용할 부분*

이제야 좀 혼란이 왔던 부분이 잡혀진 것 같습니다!

감사합니다

**Reply by jongfeel** [reply]
문법적인 interface로 이해해도 상관 없습니다.
